src/once.rs

```diff
@@ -0,0 +1,250 @@
+//! A wait-free alternative to `std::sync::OnceLock`, with helper methods that make it easier to
+//! correctly register state at startup.
+use std::{error::Error, fmt::Display, ptr::null_mut};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+use crate::{
+    atomic_try_update,
+    bits::{Align8, FlagPtr},
+    Atom,
+};
+
+#[derive(IntoPrimitive, TryFromPrimitive)]
+#[repr(usize)]
+enum Lifecycle {
+    NotSet = 0,
+    Setting,
+    Set,
+    Dead,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum OnceLockFreeError {
+    AlreadySet,
+    AttemptToReadWhenUnset,
+    AttemptToSetConcurrently,
+    InShutdown,
```

Comment:
The first three feel pretty clear. From reading the code, I interpret `InShutdown` as to mean this entity is already end-of-life'd (I didn't dig into `UnpreparedForSet`). I really want to propose a different term for `InShutdown`, but I'm honestly drawing a blank. At a minimum it would probably worth adding a short comment for each about what they imply.

Reply:
I'm thinking about turning the last one into `UseAfterFreeBug`